### PR TITLE
feat(anonymizer): provider-aware streaming deanonymizer interface [closes #52]

### LINKED ISSUE #52: feat: provider-aware streaming deanonymizer interface
## Problem

The current `StreamingDeanonymize` assumes Anthropic SSE format (`content_block_delta`, `text_delta`, `thinking_delta`). As other providers are added (OpenAI, Gemini, etc.) their SSE schemas differ and the current implementation cannot handle them correctly.

## Proposal

Define a `StreamingDeanonymizer` interface with a provider-specific implementation per AI backend, each with its own test suite.

```go
type StreamingDeanonymizer interface {
    Deanonymize(src io.ReadCloser, sessionID string) io.ReadCloser
}
```

Implementations:
- `AnthropicStreamingDeanonymizer` — current logic (content_block_delta, text_delta, thinking_delta)
- `OpenAIStreamingDeanonymizer` — OpenAI/compatible SSE format
- `GeminiStreamingDeanonymizer` — Gemini SSE format

Each implementation requires its own unit test suite covering token splits, EOF flush, and provider-specific event types.

## Acceptance Criteria

- [ ] Interface defined in `internal/anonymizer`
- [ ] Anthropic implementation extracted from current code
- [ ] Each provider implementation has test coverage >= 85%
- [ ] Provider selected based on request routing context

## Notes

This should be addressed before any new provider is integrated. Opened during review of PR #51.

### PR DESCRIPTION
**Pull request by iamclaude697**
## What

Extracts a `StreamingDeanonymizer` interface so that each AI API provider's SSE format is handled by a dedicated implementation. Previously, `StreamingDeanonymize` hardcoded Anthropic's `content_block_delta` format — non-Anthropic providers fell through to raw string replacement, losing the accumulator-based token-split protection.

## Changes

**Interface + registry** (`streaming_provider.go`):
- `StreamingDeanonymizer` interface: `ProcessDataPayload(payload []byte) bool` + `Flush()`
- `ProviderForDomain(domain)` maps 9 registered domains to 6 provider types
- Factory `NewStreamingDeanonymizer` creates the appropriate implementation

**Provider implementations** (one file each):
- `streaming_anthropic.go` — extracted from `streaming.go` (mechanical move)
- `streaming_openai.go` — `choices[0].delta.content` (covers OpenAI, Mistral, Together, Perplexity, Hugging Face)
- `streaming_gemini.go` — `candidates[0].content.parts[0].text`
- `streaming_cohere.go` — `content-delta` event type with `delta.message.content.text`
- `streaming_replicate.go` — plain text SSE (not JSON)
- `streaming_passthrough.go` — raw replacement fallback for unknown domains

**Framework refactor** (`streaming.go`):
- `streamContext` simplified: removed accumulator fields, added `provider StreamingDeanonymizer`
- `processLine` reduced to ~20 lines — delegates `data:` payloads to provider
- `handleStreamEnd` calls `provider.Flush()` instead of hardcoded flush functions
- Added `writePipe` helper for proper pipe write error handling (stop on first error)

**Proxy threading** (`proxy.go`):
- `domain` parameter threaded through `deanonymizeResponseBody` → `StreamingDeanonymize`
- Both MITM and HTTP paths pass the domain from their existing context

**Documentation** updated: `docs/anonymizer.md` and `docs/architecture.md` reflect multi-provider architecture.

## Quality Gates

- [x] `go build ./...` passes
- [x] `go vet ./...` passes
- [x] All existing streaming tests pass (35+ Anthropic tests unchanged)
- [x] All new provider tests pass (26 tests across 6 files)
- [x] Coverage: 97.4% overall on `internal/anonymizer`
- [x] Delta coverage: all new files above 90%
- [x] Zero `nolint:errcheck` in new provider files (uses `writePipe` helper)
- [x] Architecture invariants maintained

## Test Plan

- Anthropic: existing 35+ tests via `readStreamResult` (domain=`api.anthropic.com`)
- OpenAI: token split, EOF flush, `[DONE]` sentinel, role-only chunk, finish_reason, verbose logging
- Gemini: token split, EOF flush, empty candidates
- Cohere: token split, EOF flush, non-content event flush
- Replicate: token split, EOF flush, done event, verbose logging
- Passthrough: raw replacement, no-accumulation verification
- Provider registry: all 9 domains + unknown + empty + localhost

## Linked Issues

Closes #52
Related: #102 (nolint:errcheck cleanup for existing code — opened during this PR)

### CODE REVIEW

File: internal/anonymizer/streaming_cohere.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+		return true
+	}
+
+	toReplace := accumulated[:flushUpTo]
+	replaced := c.opts.replacer.Replace(toReplace)
+	if toReplace != replaced && c.opts.verbose {
+		log.Printf("[DEANON] cohere text replaced: sessionID=%s tokens=%d", c.opts.sessionID, c.opts.tokenCount)
+	}
+
+	envelope.Delta.Message.Content.Text = replaced
```
Comment (iamclaude697) [inline internal/anonymizer/streaming_cohere.go:84]: **MEDIUM — accumulator state corruption on unreachable error path**

`json.Marshal` on this struct (only string/int fields) cannot fail, but the error handling is incorrect: at this point `c.textAccum` already holds the new text (written at line 68). Returning `false` tells the framework to do raw replacement on the *original* payload, but the accumulator is not rolled back. On the next event, the accumulated text from this event is counted twice — once in the accumulator, once via the framework's raw fallback.

The Anthropic provider correctly handles this with `newPayload, _ := json.Marshal(envelope)` and a comment "error impossible: only string/int fields." Recommend aligning all three (Cohere, Gemini, OpenAI) to the same pattern:

```go
newPayload, _ := json.Marshal(envelope) // error impossible: only string/int fields
```

Same issue in `streaming_gemini.go` and `streaming_openai.go`.

File: internal/anonymizer/streaming_replicate.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+		writePipe(r.opts.pw, []byte(sseDataPrefix), payload, []byte("\n"))
+		return true
+	}
+
+	r.textAccum.WriteString(string(payload))
+	accumulated := r.textAccum.String()
+
+	flushUpTo := safeCutPoint(accumulated)
+	if flushUpTo == 0 {
+		return true
```
Comment (iamclaude697) [inline internal/anonymizer/streaming_replicate.go:38]: **LOW — raw payload accumulated, but trimmed version checked**

`strings.TrimSpace` is used above (line 27) to check for `{}` or empty, but the accumulator here receives the untrimmed `string(payload)`. If a Replicate payload arrives with leading/trailing whitespace (e.g., `  hello  `), the whitespace is accumulated and could shift `safeCutPoint` boundaries.

In practice Replicate payloads are tight so this is unlikely to surface, but for consistency with the trim check:

```go
r.textAccum.WriteString(text)
```

(reuse the already-trimmed `text` variable from line 27)

File: internal/anonymizer/streaming_passthrough_test.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+// TestPassthroughFlushIsNoOp verifies that calling Flush on a
+// passthroughDeanonymizer neither panics nor writes anything to the pipe.
+// The passthrough provider does not accumulate, so Flush is intentionally
+// a no-op.
+func TestPassthroughFlushIsNoOp(t *testing.T) {
+	got := readStreamResultForDomain(t, "data: hello\n\n", nil, unknownDomain)
+	// Just verify no panic — empty token map means no replacements.
+	if got == "" {
+		t.Error("expected non-empty output from passthrough")
+	}
```
Comment (iamclaude697) [inline internal/anonymizer/streaming_passthrough_test.go:43]: **LOW — likely lint failure: missing blank line + orphaned comment**

This comment is separated from the function below by only a blank line, but there's no blank line between the closing `}` of `TestPassthroughFlushIsNoOp` above and this comment. Also, the comment doesn't start with the function name (`TestPassthroughStreamingNoAccumulation`), making it an orphaned comment rather than a Go doc comment.

This pattern — `}\n// comment\nfunc Test...` — is commonly flagged by golangci-lint. There's a similar instance in `streaming_replicate_test.go` before `TestReplicateStreamingDoneEvent` (line ~67). Both may be contributing to the CI lint failure.

**Review by iamclaude697 — COMMENTED** [review]
## PR Review: feat(anonymizer): provider-aware streaming deanonymizer interface [closes #52]

### Pipeline status
**BLOCKED** — Lint fails, Build skipped. CI log was unreachable (GitHub timeout), but two formatting issues in test files (`streaming_passthrough_test.go`, `streaming_replicate_test.go`) are likely contributors — see inline comments.

### Architecture assessment

Clean strategy pattern extraction. The `StreamingDeanonymizer` interface with `ProcessDataPayload`/`Flush` is well-designed, and the domain→provider registry is straightforward. The `writePipe` helper is a good improvement over scattered `//nolint:errcheck` in provider files. Documentation updates are thorough.

The refactor preserves Anthropic behavioral parity — non-delta events trigger flush before passthrough, EOF calls `provider.Flush()`, and the framework's raw-replacement fallback handles unparseable payloads identically to the old `json.Unmarshal` error path.

### Test coverage verdict
**PASS** — 26 new tests across 6 provider test files. Token-split, EOF-flush, verbose-logging, invalid-JSON, and edge-case paths are all covered. The passthrough limitation test (`TestPassthroughStreamingNoAccumulation`) is a particularly good pattern — testing documented trade-offs.

### Findings summary

| Severity | File | Issue |
|----------|------|-------|
| MEDIUM | `streaming_cohere.go`, `streaming_gemini.go`, `streaming_openai.go` | `json.Marshal` error path returns `false` after accumulator mutation — theoretically corrupts state. Anthropic uses `_` correctly. Align all providers. |
| LOW | `streaming_replicate.go` | `textAccum.WriteString(string(payload))` accumulates untrimmed bytes but `{}` check uses `TrimSpace`. Use the already-trimmed variable. |
| LOW | `streaming_passthrough_test.go`, `streaming_replicate_test.go` | Missing blank lines / orphaned comments before test functions — likely lint violations. |

### Path to Approval

1. Fix lint failures (test file formatting, and any others revealed once CI logs are accessible)
2. Unify `json.Marshal` error handling across Cohere/Gemini/OpenAI to match Anthropic's `_` pattern (3 one-line changes)
3. Optional: use trimmed `text` variable in Replicate accumulator

### Verdict
**Not ready** — blocked on CI. Code quality is high; the findings above are minor. Once lint passes and the `json.Marshal` pattern is unified, this is ready to merge.

**Review by iamclaude697 — COMMENTED** [review]
## Updated: Pipeline is green

CI lint failure has been resolved — all checks pass now.

### Revised Path to Approval

1. Unify `json.Marshal` error handling across Cohere/Gemini/OpenAI to match Anthropic's `_` pattern (3 one-line changes) — see inline comment on `streaming_cohere.go`
2. Optional: use trimmed `text` variable in Replicate accumulator

### Revised Verdict
**Nearly ready** — one MEDIUM finding remains (the `json.Marshal` error pattern). Once that's addressed, this is good to merge.

File: internal/anonymizer/streaming_cohere.go (hunk 1 above)
Comment (iamclaude697) [reply internal/anonymizer/streaming_cohere.go:84]: Fixed in 516c6a5. Aligned all three (Cohere, Gemini, OpenAI) to the Anthropic pattern: `newPayload, _ := json.Marshal(envelope)` with the "error impossible" comment. The `return false` path was indeed wrong — accumulator already mutated, so the framework's raw fallback would double-count.

File: internal/anonymizer/streaming_replicate.go (hunk 2 above)
Comment (iamclaude697) [reply internal/anonymizer/streaming_replicate.go:38]: Fixed in 516c6a5. Now accumulates `text` (the trimmed variable) instead of `string(payload)`.

File: internal/anonymizer/streaming_passthrough_test.go (hunk 3 above)
Comment (iamclaude697) [reply internal/anonymizer/streaming_passthrough_test.go:43]: Fixed in 516c6a5. Both comments now start with the function name (`TestPassthroughStreamingNoAccumulation`, `TestReplicateStreamingDoneEvent`). The CI lint failure was actually gofmt + misspell (fixed in 50d8b86), not the orphaned comments — but the doc comment style was still wrong, so fixed both.

**Review by iamclaude697 — COMMENTED** [review]
All findings addressed. CI is green across all 8 checks. Clean to merge. (Can't formally APPROVE own PR — this serves as the review sign-off.)

**Review by laplaque — APPROVED** [review]
All findings addressed. CI is green across all 8 checks. Clean to merge.
